platform/linux-generic/include/odp_chksum_internal.h

```diff
@@ -0,0 +1,78 @@
+/* Copyright (c) 2018, Linaro Limited
+ * All rights reserved.
+ *
+ * SPDX-License-Identifier:     BSD-3-Clause
+ */
+
+#ifndef ODP_CHKSUM_INTERNAL_H_
+#define ODP_CHKSUM_INTERNAL_H_
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+#include <odp/api/abi/byteorder.h>
+
+/* Simple implementation of ones complement sum.
+ * Based on RFC1071 and its errata.
+ */
+typedef union {
+	uint16_t w;
+	uint8_t b[2];
+} swap_buf_t;
+
+static uint32_t _odp_chksum_ones_comp16_32(const uint8_t *p, uint32_t len,
+					   odp_bool_t odd_offset)
+
```

Comment:
odd_offset and p%2 seem overlapping checks. Couldn't this function always check and move to an even address and then use the API function for the rest of the (now even address) calculation, and combine results in the end.

Pseudo code:

```
uint32_t odd_checksum(uint8_t *p, uint32_t len)
{
    if (p%2) {
        sw.b[0] = 0;
        sw.b[1] = *(p++);
        odd_sum = sw.w;
        len--;
    }

    even_sum = odp_chksum_ones_comp16(p, len);

    return odd_sum + even_sum;
}
```

Reply:
@psavol Maybe I should add this as an in-source comment. There is difference between in-packet offset oddity and data address offset oddity. As I said, I'll think about improving this low-level function. Compare it with helper function, which also has this odd/even in-packet support, but always uses `uint16_t` to further access data.